druid-shell/src/backend/wayland/display.rs

```diff
@@ -0,0 +1,228 @@
+        self.subscriptions.borrow_mut().insert(sub.id, sub.clone());
+        sub
+    }
+
+    fn release(&self, s: &GlobalEventSubscription) {
+        self.subscriptions.borrow_mut().remove(&s.id);
+    }
+}
+
+#[allow(unused)]
+pub(super) struct Environment {
+    pub(super) display: wlc::Display,
+    pub(super) registry: wlc::GlobalManager,
+    pub(super) xdg_base: wlc::Main<xdg_wm_base::XdgWmBase>,
+    pub(super) queue: std::rc::Rc<std::cell::RefCell<wlc::EventQueue>>,
+    dispatcher: std::sync::Arc<Dispatcher>,
+}
+
+unsafe impl Sync for Environment {}
+unsafe impl Send for Environment {}
+
+impl GlobalEventDispatch for std::sync::Arc<Environment> {
+    fn subscribe(&self, sub: impl Into<GlobalEventSubscription>) -> GlobalEventSubscription {
+        self.dispatcher.subscribe(sub)
+    }
+
+    fn release(&self, s: &GlobalEventSubscription) {
+        self.dispatcher.release(s)
+    }
+}
+
+pub(super) fn new(dispatcher: Dispatcher) -> Result<std::sync::Arc<Environment>, error::Error> {
+    let dispatcher = std::sync::Arc::new(dispatcher);
+    let d = wlc::Display::connect_to_env()?;
+
+    let mut queue = d.create_event_queue();
+    let handle = d.clone().attach(queue.token());
+    let registry = wlc::GlobalManager::new_with_cb(&handle, {
+        let dispatcher = dispatcher.clone();
+        move |event, registry, ctx| {
+            dispatcher.consume(&event, &registry, &ctx);
+        }
+    });
+
+    // do a round trip to make sure we have all the globals
+    queue
+        .sync_roundtrip(&mut (), |_, _, _| unreachable!())
+        .map_err(error::Error::fatal)?;
+
+    let xdg_base = registry
+        .instantiate_exact::<xdg_wm_base::XdgWmBase>(2)
+        .map_err(|e| error::Error::global("xdg_wm_base", 2, e))?;
+
+    // We do this to make sure wayland knows we're still responsive.
+    //
+    // NOTE: This means that clients mustn't hold up the event loop, or else wayland might kill
+    // your app's connection. Move *everything* to another thread, including e.g. file i/o,
+    // computation, network, ... This is good practice for all back-ends: it will improve
+    // responsiveness.
+    xdg_base.quick_assign(|xdg_base, event, ctx| {
+        tracing::info!(
+            "global xdg_base events {:?} {:?} {:?}",
+            xdg_base,
+            event,
+            ctx
+        );
+        match event {
+            xdg_wm_base::Event::Ping { serial } => xdg_base.pong(serial),
+            _ => (),
+        }
+    });
+
+    let env = std::sync::Arc::new(Environment {
+        queue: std::rc::Rc::new(std::cell::RefCell::new(queue)),
+        display: d,
+        registry,
+        xdg_base,
+        dispatcher,
+    });
+
+    return Ok(env);
+}
+
+pub(super) fn global() -> Result<std::sync::Arc<Environment>, error::Error> {
+    use lazy_static::lazy_static;
+    lazy_static! {
+        static ref _GLOBAL: std::sync::Mutex<Option<std::cell::RefCell<std::sync::Arc<Environment>>>> =
```

Comment:
`Mutex<Option<RefCell<Arc<>>>>` feels very weird, `Mutex<Option<Arc<>>>` should just work.

Reply:
i remember this particular piece being a PITA to determine. will take another look now that the code is working and stabilized though.

Reply:
also if you want a "global storage" you can create these in `Application::new()` and store it in `Application`, then you can access the global data using `druid_shell::Application::global().0.data`

Reply:
indeed, thought about it, dislike reaching outside the module and making that particular assumption for this. either screens is global and doesn't depending on applications or its not and it should take a reference of some sort in its function args.